addons/binding/org.openhab.binding.mysensors/src/main/java/org/openhab/binding/mysensors/handler/MySensorsThingHandler.java

```diff
@@ -0,0 +1,396 @@
+
+                        newMsg.setSubType(subType);
+                        newMsg.setMsg(adapter.fromCommand(command));
+
+                        myGateway.sendMessage(newMsg);
+
+                    } else {
+                        logger.warn("Variable not found, cannot handle command for thing {} of type {}", thing.getUID(),
+                                channelUID.getId());
+                    }
+                } else {
+                    logger.error("Could not get type of variable for channel: {}, command: {} of thing {}",
+                            thing.getUID(), command, thing.getUID());
+                }
+
+            } else {
+                logger.error("Type adapter not found for {}", channelUID.getId());
+            }
+
+        }
+    }
+
+    @Override
+    public void messageReceived(MySensorsMessage message) throws Exception {
+        handleIncomingMessageEvent(message);
+
+    }
+
+    @Override
+    public void sensorUpdateEvent(MySensorsNode node, MySensorsChild child, MySensorsVariable var,
+            MySensorsNodeUpdateEventType eventType) {
+        switch (eventType) {
+            case UPDATE:
+            case REVERT:
+                if ((node.getNodeId() == nodeId) && (child.getChildId() == childId)) {
+                    handleChildUpdateEvent(var);
+                    updateLastUpdate(node, eventType == MySensorsNodeUpdateEventType.REVERT);
+                }
+                break;
+            case BATTERY:
+                if (node.getNodeId() == nodeId) {
+                    handleBatteryUpdateEvent(node);
+                    updateLastUpdate(node, eventType == MySensorsNodeUpdateEventType.REVERT);
+                }
+                break;
+        }
+
+    }
+
+    @Override
+    public void connectionStatusUpdate(MySensorsAbstractConnection connection, boolean connected) throws Exception {
+        if (!connected) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+        } else {
+            updateStatus(ThingStatus.ONLINE);
+        }
+    }
+
+    @Override
+    public void nodeReachStatusChanged(MySensorsNode node, boolean reach) {
+        if (node.getNodeId() == nodeId) {
+            if (!reach) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            } else {
+                updateStatus(ThingStatus.ONLINE);
+            }
+        }
+
+    }
+
+    /**
+     * For every thing there is a lastUpdate channel in which the date/time is stored
+     * a message was received from this thing.
+     */
+    private void updateLastUpdate(MySensorsNode node, boolean isRevert) {
+        // Don't always fire last update channel, do it only after a minute by
+        if (lastUpdate == null || (System.currentTimeMillis() > (lastUpdate.getCalendar().getTimeInMillis() + 60000))
+                || revertState) {

```

Comment:
Short explanation of the mechanism: https://github.com/tobof/openhab2-addons/wiki/Configuration#enable-ack-request-for-things

boolean revertState: should the binding try to revert the status if no ACK was received?
boolean isRevert: is the current update of the thing a revert (process)

Example:
1) A switch is toggled from OFF to ON
2) the binding sends ON to the node with the flag ACK(noledgement) set to 1 (true).
3a) the nodes receives the ON, answers with an ACK and everything is fine. (END)
3b) the nodes does not respond to the command
4) the binding will try to resend the command five times
5) if no answer was received the binding will check if an old status (OFF) is known
6) if there is a known old status it will try to revert the state of the switch (isRevert == true)